### DISCUSSION TITLE
Enhance: Translation Option Strings

### DISCUSSION THREAD

**Discussion by flatcap** [neomutt/neomutt#4002] [Development]
This is a **coding** task; you don't need to be multi-lingual :-)
It's a little hard to explain, but the coding should be simpler.

## Current Translation Mechanism

All the strings shown to the user are given to `gettext(3)` to be translated.

When it comes to Menus, this is a little trickier.
There are two strings which must be kepts in sync:

**English**:
- Message: "Sort by (**d**)ate, (**a**)lpha, si(**z**)e?"
- Option letters: "**daz**"

**German**:
- Message: "Sortiere nach da(**t**)um, a(**l**)phabet, (**g**)röße?"
- Option letters: "**tlg**"

A German user is presented with the message and NeoMutt expects <kbd>t</kbd>, <kbd>l</kbd> or <kbd>g</kbd> in response.
NeoMutt looks up the letter's index in the **translated** option letters. 
This index is used in the code

In `browser.c`:
```c
case 1: /* (d)ate */
  sort = SORT_DATE;
  break;

case 2: /* (a)lpha */
  sort = SORT_SUBJECT;
  break;

case 3: /* si(z)e */
  sort = SORT_SIZE;
  break;
```

The "option letters" string allows an important flexibility.
The translator can reorder the message to fit their language, as long as the
"option letters" string stays in the same order as the English version.

**English**:
- Message: "(**A**)pple, (**B**)anana, (**C**)herry"
- Option letters: "**abc**"

**German**:
- Message: "(**K**)irsche, (**B**)anane, (**A**)pfel"
- Option letters: "**abk**"

The German **messages** are in reverse order, but the **option letters** are in the English order.

## A Solution

NeoMutt needs to know the index in the English order, so we need to encode that into the message.
It would look like:

**English**, **German**:
- Message: "(**A**)pple, (**B**)anana, (**C**)herry"
- Message: "(**3:K**)irsche, (**2:B**)anane, (**1:A**)pfel"

The parenthesised letter is given the index number from the English message.
- The number is **optional**
- The number may be two digits

If the translated message is in the same order as the English message,
then the number can be omitted:

**English**, **German**:
- Message: "(**A**)pple, (**B**)anana, (**C**)herry"
- Message: "(**A**)pfel, (**B**)anane, (**K**)irsche"

## How Will That Work?

The menu function will parse the string looking for `(x)` or `(d:x)` patterns.

Currently:
- All the messages are in this format
- All the translations are in this format
- All the translations are in the same order as the English

It will generate a lookup table of index:letter.

If there are duplicate, or missing, index numbers then the English message will be used (and the failure logged).

This means that all the callers need changing to use `N_()` gettext macros.

- `mutt_multi_choice(_("(m)essage"));`
to
- `mutt_multi_choice(N_("(m)essage"));`

## Task

- Drop the `letters` param from `mutt_multi_choice()`

- Change the `messages` param from `_("...")` to `N_("...")`
  This means `gettext()` will need to be called inside `mutt_multi_choice()`

- Create a function to parse the new message format into an index:letter array

- Alter `mutt_multi_choice()` to parse the translated message, or fallback to the English message.
  The message can be manually translated by calling `gettext()`.

**Comment by abc123098123** [comment]
@flatcap 

Hey!

I am interested in solving this issue. Still have some doubts about it. But I'll start with some obvious things, then I will ask you some general questions about it.

Thanks!

**Comment by flatcap** [comment]
I'd start with a commit adding some test code:
```c
mutt_multi_choice(_("(a)pple, (b)anana, (c)herry"), _("abc"));
goto main_ok;
```

just before

https://github.com/neomutt/neomutt/blob/9c0d566dd71b5a19d0e2280ff9bfa53a631fbe7f/main.c#L812-L815

No point in building/running the whole of NeoMutt.

I'll try to create a list of test cases.
Also, we need to find out how to use locally compiled `.mo` files, rather than the system ones.

**Comment by flatcap** [comment]
If you make changes to the `.po` files, you can test them by:

- Install NeoMutt to `~/install`
  `make DESTDIR=$HOME/install install`

- Set the text domain
  `export TEXTDOMAINDIR="$HOME/install/usr/share/locale"`

- Run NeoMutt using the updated language
  `export LANG=de_DE.utf-8`
  `./neomutt`

**Comment by flatcap** [comment]
## Data

- "Apple", "Banana", "Cherry"
- "Apfel", "Banane", "Kirsche"

## Results

`mutt_multi_choice()` will return a number between `1` and `n`.

- 1 display "APPLE"
- 2 display "BANANA"
- 3 display "CHERRY"

## Simple Parsing Tests

Parsing the following strings will produce:

| String                              | Lookup table  |
| :---------------------------------- | :------------ |
| "(a)pple, (b)anana, (c)herry"       | a=1, b=2, c=3 |
| "ap(p)le, ba(n)ana, ch(e)rry"       | p=1, n=2, e=3 |
| "appl(e), banan(a), cherr(y)"       | e=1, a=2, y=3 |
| "(1) apple, (2) banana, (3) cherry" | [NONE]        |

## Sequence Parsing Tests

Parsing these sequence strings will produce:

| String                               | Lookup table  |
| :----------------------------------- | :------------ |
| "Ap(f)el, Ba(n)ane, Ki(r)sche"       | f=1, n=2, r=3 |
| "Ap(1:f)el, Ba(2:n)ane, Ki(3:r)sche" | f=1, n=2, r=3 |
| "Ap(1:f)el, Ki(3:r)sche, Ba(2:n)ane" | f=1, r=3, n=2 |
| "Ba(2:n)ane, Ap(1:f)el, Ki(3:r)sche" | n=2, f=1, r=3 |
| "Ba(2:n)ane, Ki(3:r)sche, Ap(1:f)el" | n=2, r=3, f=1 |
| "Ki(3:r)sche, Ap(1:f)el, Ba(2:n)ane" | r=3, f=1, n=2 |
| "Ki(3:r)sche, Ba(2:n)ane, Ap(1:f)el" | r=3, n=2, f=1 |

## Key Press Tests

Here are some sample strings and the results of pressing keys.

### Simple

- "(a)pple, (b)anana, (c)herry"

| Key          | Result   |
| :----------- | :------- |
| <kbd>a</kbd> | "APPLE"  |
| <kbd>1</kbd> | "APPLE"  |
| <kbd>b</kbd> | "BANANA" |
| <kbd>2</kbd> | "BANANA" |
| <kbd>c</kbd> | "CHERRY" |
| <kbd>3</kbd> | "CHERRY" |

### Simple Translation

- "Ap(f)el, Ba(n)ane, Ki(r)sche"

| Key          | Result   |
| :----------- | :------- |
| <kbd>f</kbd> | "APPLE"  |
| <kbd>1</kbd> | "APPLE"  |
| <kbd>n</kbd> | "BANANA" |
| <kbd>2</kbd> | "BANANA" |
| <kbd>r</kbd> | "CHERRY" |
| <kbd>3</kbd> | "CHERRY" |

### Sequence Translation

- "Ap(1:f)el, Ba(2:n)ane, Ki(3:r)sche"

| Key          | Result   |
| :----------- | :------- |
| <kbd>f</kbd> | "APPLE"  |
| <kbd>1</kbd> | "APPLE"  |
| <kbd>n</kbd> | "BANANA" |
| <kbd>2</kbd> | "BANANA" |
| <kbd>r</kbd> | "CHERRY" |
| <kbd>3</kbd> | "CHERRY" |

### Sequence Out-of-order Translation

- "Ki(3:r)sche, Ap(1:f)el, Ba(2:n)ane"

| Key          | Result   |
| :----------- | :------- |
| <kbd>f</kbd> | "APPLE"  |
| <kbd>1</kbd> | "APPLE"  |
| <kbd>n</kbd> | "BANANA" |
| <kbd>2</kbd> | "BANANA" |
| <kbd>r</kbd> | "CHERRY" |
| <kbd>3</kbd> | "CHERRY" |

## Bad strings

- no `()`s
- empty `()`s
- multiple chars in `()`
- missing param `(:X)` or `(3:)`
- missing seq `(1:a)`, `(3:c)`
- no translation

## Notes

- accented letters?
- ignore case?

**Comment by abc123098123** [comment]
Hi @flatcap !

I am sorry, I have been a bit busy this week, but I am ready to tackle this issue.
I have some questions:

1. What is the actual workflow for this? I know I have to change code (parsing `(d:x)`). But what about the rest? It seems I have to change `po` files (translation and addition of the `(d:x)` format)?
2. By testing the `.po` files like you said (installing on `$HOME/install`), shouldn't I be getting german inside Neomutt? (Everything is in english).

Thank you very much!

**Comment by flatcap** [comment]
> I have been a bit busy this week

no worries

> By testing the `.po` files like you said (installing on `$HOME/install`),
> shouldn't I be getting german inside Neomutt? (Everything is in english).

1. `make DESTDIR=$HOME/install install`
   Puts the files in the right tree structure
2. `export TEXTDOMAINDIR="$HOME/install/usr/share/locale"`
   Tells NeoMutt where to look for the translations
3. `export LANG=de_DE.utf-8`  (note, I've added `.utf-8`)
   Tells NeoMutt which language to use

I checked `neomutt -v` and the help page inside NeoMutt.
German is 92% complete, so most of NeoMutt should be changed.  It is for me.

> What is the actual workflow for this?

The function will:
- Take an English string
- Translate it (call `gettext()`)
- Parse the option letters into a lookup table
- Display the question
- Lookup the letter in the table
- Return the response

If the translation isn't valid, we'll fallback to using the English string.

> It seems I have to change `po` files

Nope!  That's the beauty of it.
When we came up with the plan, we realised that it needed to be backwards-compatible.

Once finished, your code will:
1. Accept `(a)pple, (b)anana, (c)herry`
   It took make a LONG time to massage all the strings and all thirty translations into this format.
  You won't need to alter any `.po` files.

2. Eliminate the `'abc'` letter strings
   This will hugely simplify the maintenance of options.
   You don't need to delete them, as the update process will automatically mark them as dead.

3. Allow translators to reorder the options to suit their language.
   Once we document the new format, the translators can adopt it.

**Comment by abc123098123** [comment]
Thank you so much. I think I have enough to start with.

But I still continue to not get German in my neomutt. It's a bit late, maybe I am doing something wrong.

My configure command:
```
./configure \
        --prefix=/usr \
        --sysconfdir=/etc \
        --libexecdir=/usr/lib \
        --gpgme \
        --sqlite \
        --autocrypt \
        --lua \
        --notmuch \
        --gss \
        --gnutls \
        --sasl \
        --with-ui=ncurses \
        --with-idn2=/usr \
        --disable-idn \
        --idn2 \
        --lmdb \
        --kyotocabinet \
        --gdbm \
        --lz4 \
        --zlib \
        --zstd
```

Then I do the three steps you mentioned. I go to help page and it is all in English still.
This is my `neomutt -v`:

```
NeoMutt 20200925-51-9c0d566-dirty
Copyright (C) 1996-2020 Michael R. Elkins and others.
NeoMutt comes with ABSOLUTELY NO WARRANTY; for details type 'neomutt -vv'.
NeoMutt is free software, and you are welcome to redistribute it
under certain conditions; type 'neomutt -vv' for details.

System: Linux 5.8.2-zen1-1-zen (x86_64)
ncurses: ncurses 6.2.20200212 (compiled with 6.2.20200212)
libidn2: 2.3.0 (compiled with 2.3.0)
GPGME: 1.14.0
GnuTLS: 3.6.14
libnotmuch: 5.2.0
storage: kyotocabinet, gdbm, lmdb
compression: lz4, zlib, zstd

Configure options: --prefix=/usr --sysconfdir=/etc --libexecdir=/usr/lib --gpgme --sqlite --autocrypt --lua --notmuch --gss --gnutls --sasl --with-ui=ncurses --with-idn2=/usr --disable-idn --idn2 --lmdb --kyotocabinet --gdbm --lz4 --zlib --zstd

Compilation CFLAGS: -g -O2 -std=c99 -fno-delete-null-pointer-checks -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -I/usr/include -DNCURSES_WIDECHAR

Default options:
  +attach_headers_color +compose_to_sender +compress +cond_date +debug 
  +encrypt_to_self +forgotten_attachments +forwref +ifdef +imap +index_color 
  +initials +limit_current_thread +multiple_fcc +nested_if +new_mail +nntp +pop 
  +progress +quasi_delete +regcomp +reply_with_xorig +sensible_browser +sidebar 
  +skip_quoted +smtp +status_color +timeout +tls_sni +trash 

Compile options:
  +autocrypt +bkgdset +color +curs_set +fcntl -flock -fmemopen +futimens 
  +getaddrinfo +gnutls +gpgme +gss +hcache -homespool +idn +inotify 
  -locales_hack +lua +meta -mixmaster +nls +notmuch -openssl +pgp +regex +sasl 
  +smime +sqlite +start_color +sun_attachment +typeahead 
MAILPATH="/var/mail"
PKGDATADIR="/usr/share/neomutt"
SENDMAIL="/usr/sbin/sendmail"
SYSCONFDIR="/etc"

To learn more about NeoMutt, visit: https://neomutt.org
If you find a bug in NeoMutt, please raise an issue at:
    https://github.com/neomutt/neomutt/issues
or send an email to: <neomutt-devel@neomutt.org>
```

Just like you said, it should be in German, if all steps done correctly, right?

I wonder if its something related to the `locale` system-wide?

Thanks!

**Comment by flatcap** [comment]
Hmm...
I've just tried your exact `./configure`.
I wondered if the prefix options might be causing trouble, but they're not.
I capitalised a section of the German `.po` file, so I'm definitely seeing that one and not the system-installed version.

**Comment by abc123098123** [comment]
Could it be my `/etc/locale.conf` ?
I only have one line in there and its `LANG=en_US.UTF-8`.

Does `/etc/locale.gen` have to have `de_DE.UTF-8` uncommented?

**Comment by flatcap** [comment]
> Could it be my `/etc/locale.conf` ?
> I only have one line in there and its `LANG=en_US.UTF-8`.

I don't think so.  I think that's the machine default; your environment should overrule _that_.

Which shell are you using (I'm using `bash`, some need `setenv` rather than `export`)

**Comment by abc123098123** [comment]
I am using `zsh`. When I do `export` I always confirm if its set or not (I confirm it is).

When I export `LANG` to German, this is what happens (only an example):
![image](https://user-images.githubusercontent.com/2325589/95689130-b28f3980-0c06-11eb-93f1-61a875c40156.png)

While this is what should be shown (but in German):
![image](https://user-images.githubusercontent.com/2325589/95689149-cdfa4480-0c06-11eb-95d2-aa892b69aa17.png)

I also tried in `bash`, but the results are the same, unfortunately.

**Comment by flatcap** [comment]
OK, the junk chars are caused by a non-utf8 locale,
but the rest of the text won't be changed because surely that's your `$status_format` setting?

The other thing I tested was the "sub-lang".
It needed to be `de_DE.utf-8`, plain `de.utf-8` didn't work.

**Comment by abc123098123** [comment]
My current `$status_format` is:
```
set status_format = "───[ Folder: %f ]───[%r%m messages%?n? (%n new)?%?d? (%d to delete)?%?t? (%t tagged)? ]───%>─%?p?( %p postponed )?───"
```

Also, I'm sorry, but what do you mean by sub-lang? Where do you change it?

**Comment by flatcap** [comment]
The status_format won't be translated, but the unicode `───` chars just got mangled (because neomutt didn't know it was utf-8).

> sub-lang

it's the `_XX` bit.
`LANG=en` (general English)
`LANG=en_GB` (**British** English)

I'm not sure why not, but NeoMutt didn't pick up the translation unless I used the exact string:
`export LANG=de_DE.utf-8`

**Comment by abc123098123** [comment]
Well, it seems I've just hit a brick wall then... :-(

**Comment by flatcap** [comment]
OK, two last ideas.  `strace` and brute force.

```sh
export LANG=de_DE.utf-8
strace -o s1 neomutt -v
grep neomutt.mo s1
```

I see this, as NeoMutt looks for the translation:

```
openat(AT_FDCWD, "/usr/share/locale/de_DE.utf-8/LC_MESSAGES/neomutt.mo", O_RDONLY) = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
openat(AT_FDCWD, "/usr/share/locale/de_DE.utf8/LC_MESSAGES/neomutt.mo", O_RDONLY) = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
openat(AT_FDCWD, "/usr/share/locale/de_DE/LC_MESSAGES/neomutt.mo", O_RDONLY) = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
openat(AT_FDCWD, "/usr/share/locale/de.utf-8/LC_MESSAGES/neomutt.mo", O_RDONLY) = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
openat(AT_FDCWD, "/usr/share/locale/de.utf8/LC_MESSAGES/neomutt.mo", O_RDONLY) = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
openat(AT_FDCWD, "/usr/share/locale/de/LC_MESSAGES/neomutt.mo", O_RDONLY) = 3
```

Do you see these lookups?
Now try after the `export TEXTDOMAINDIR`

If you run `locale` does it give any errors?

---

My brute force method is

- Forget the install
- Forget the `export TEXTDOMAINDIR`
- As `root`:

```sh
# Go to a translation dir
cd /usr/share/locale/de/LC_MESSAGES
# If you have an installed NeoMutt:
mv neomutt.mo neomutt.mo.orig
# point at your build
ln -s $HOME/neomutt/po/de.mo neomutt.mo
```

**Comment by abc123098123** [comment]
I did the `strace` to no avail.
Nothing shows up even after `export TEXTDOMAINDIR`.

But curiously, when I run `locale`, after setting `LANG`, this happens:

```
locale: Cannot set LC_MESSAGES to default locale: No such file or directory
locale: Cannot set LC_ALL to default locale: No such file or directory
LANG=de_DE.utf-8
LC_CTYPE=en_US.UTF-8
LC_NUMERIC="de_DE.utf-8"
LC_TIME="de_DE.utf-8"
LC_COLLATE=C
LC_MONETARY="de_DE.utf-8"
LC_MESSAGES="de_DE.utf-8"
LC_PAPER="de_DE.utf-8"
LC_NAME="de_DE.utf-8"
LC_ADDRESS="de_DE.utf-8"
LC_TELEPHONE="de_DE.utf-8"
LC_MEASUREMENT="de_DE.utf-8"
LC_IDENTIFICATION="de_DE.utf-8"
LC_ALL=
```

Also, the bruteforce method didn't work. Nothing changed.

**Comment by flatcap** [comment]
> But curiously, when I run `locale`, after setting `LANG`, this happens:

That's odd.  I see exactly the opposite.

If I set to `de_DE`, locale works (and translations work)
If I set to `de`, locale gives errors (and translations don't work)

Please can you try one last time, with `export LANG=de.utf-8`
Check if `locale` is happy, first.  Then try neomutt.

If that doesn't work for you, then I'll delete this issue
(and we'll never talk of this again ;-)

**Comment by abc123098123** [comment]
This is so annoying.

I tried what you said:

```
λ  ~ export LANG="de.utf-8"   
                           
λ  ~ locale
locale: Cannot set LC_MESSAGES to default locale: No such file or directory
locale: Cannot set LC_ALL to default locale: No such file or directory
LANG=de.utf-8
LC_CTYPE=en_US.UTF-8
LC_NUMERIC="de.utf-8"
LC_TIME="de.utf-8"
LC_COLLATE=C
LC_MONETARY="de.utf-8"
LC_MESSAGES="de.utf-8"
LC_PAPER="de.utf-8"
LC_NAME="de.utf-8"
LC_ADDRESS="de.utf-8"
LC_TELEPHONE="de.utf-8"
LC_MEASUREMENT="de.utf-8"
LC_IDENTIFICATION="de.utf-8"
LC_ALL=
λ  ~ neomutt
```

Nothing different happens, except the mangled characters.
But I think its important to talk about one last thing.

When I used strace, `grep neomutt.mo s1` returned zero results. Is this normal behavior? Shouldn't neomutt fetch at least some kind of `.mo` file?

**Comment by flatcap** [comment]
I'm getting closer...
I read back through the strace output.

NeoMutt looks for some other files first:
- `/usr/lib/locale/locale-archive`
- `/usr/share/locale/locale.alias`

trying to match the `LANG` setting.
If they don't work, it then tries:
- `/usr/lib/locale/de.utf-8/LC_IDENTIFICATION`
- `/usr/lib/locale/de.utf8/LC_IDENTIFICATION`
- `/usr/lib/locale/de/LC_IDENTIFICATION`

then gives up.

**Comment by flatcap** [comment]
`localedef --list` shows all the locales that the system understands

The results **don't** include `de` only `de_*`.
That explains my results.

**Comment by abc123098123** [comment]
Hi!

Well, after doing `export LANG="de_DE.UTF-8"`, the output of `localedef --list` is as follows:
```
en_US
en_US.iso88591
en_US.utf8
```

What I don't understand is: if this displays all the locales that the system understands, shouldn't `de_DE.UTF-8` be there too for Neomutt to use the translated files?

Also, here is the `strace` relevant output:

```
openat(AT_FDCWD, "/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=3362224, ...}) = 0
mmap(NULL, 3362224, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7fd483b24000
close(3)                                = 0
openat(AT_FDCWD, "/usr/share/locale/locale.alias", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=2998, ...}) = 0
read(3, "# Locale name alias data base.\n#"..., 4096) = 2998
read(3, "", 4096)                       = 0
close(3)                                = 0
openat(AT_FDCWD, "/usr/lib/locale/de_DE.UTF-8/LC_IDENTIFICATION", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/locale/de_DE.utf8/LC_IDENTIFICATION", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/locale/de_DE/LC_IDENTIFICATION", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/locale/de.UTF-8/LC_IDENTIFICATION", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/locale/de.utf8/LC_IDENTIFICATION", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/locale/de/LC_IDENTIFICATION", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
```

It is just like you said. And it seems it has given up. Although `openat` seems to return 3?
Does this ring any bells?

**Comment by flatcap** [comment]
I've spent a couple of hours in a minimal Fedora VM messing around with `localedef`, but I'm none the wiser.
`locale-archive` didn't exist.  All the `en_*` dirs in `/usr/lib/locale/` were installed by `glibc-langpack` which comes in hundreds of language variants.

My main Fedora install only has these `en_*` dirs, but the `locale-archive` contains hundreds of languages.
I've no idea how to populate it.

I can `localedef --add-to-archive` the existing `en_*` dirs, but nothing else.

**Comment by flatcap** [comment]
hmm... installing `glibc-all-langpacks` package seems to create the archive and make things work.
`localedef --list` and `locale -a` now list lots of langs.

Aha!

Now, `LANG=de_DE.utf-8 mutt -v` translates!  (it didn't before the glibc install).
